Index: pcap-linux.c
===================================================================
--- pcap-linux.c.orig
+++ pcap-linux.c
@@ -3092,6 +3092,9 @@ activate_new(pcap_t *handle)
 			else
 				return 0;	/* try old mechanism */
 		}
+		/* Hack to make things work on s390 ctc interfaces */
+		if (strncmp("ctc", device, 3) == 0)
+			handle->linktype = DLT_EN10MB;
 	} else {
 		/*
 		 * The "any" device.
